api/src/main/kotlin/io/spine/protodata/renderer/InsertionPoint.kt

```diff
@@ -49,7 +81,93 @@ public interface InsertionPoint {
      * @return the line number at which the insertion point should be added.
      * @see LineNumber
      */
-    public fun locate(lines: List<String>): LineNumber
+    @Deprecated("Use locate(Text) instead.")
+    public fun locate(lines: List<String>): LineNumber {
+        val coords = locate(text(lines))
+        return when (coords.kindCase) {
+            WHOLE_LINE -> LineNumber.at(coords.wholeLine)
+            INLINE -> {
+                logUnsupportedKind()
+                LineNumber.at(coords.inline.cursor.line)
+            }
+            END_OF_FILE -> LineNumber.endOfFile()
+            NOT_IN_FILE -> LineNumber.notInFile()
+            else -> error("Unexpected text coordinates `$coords`.")
+        }
+    }
+
+    private fun logUnsupportedKind() =
+        loggerFor(InsertionPoint::class.java)
+            .atWarning()
+            .withStackTrace(FULL)
+            .log(
+                "`locate(List<String>)` does not support inline insertion. " +
+                        "A whole line insertion will be generated."
+            )
+
+    /**
+     * Locates the site where the insertion point should be added.
+     *
+     * An insertion point should only appear once in a file.
+     *
+     * @param text the existing code
+     * @return the coordinates in the text where the insertion point should be added
+     * @see SourceFile.at
+     * @see SourceFile.atInline
+     */
+    public fun locate(text: Text): TextCoordinates
+}
+
+/**
+ * A factory of [TextCoordinates] instances.
+ *
+ * This interface serves as a trait for the [InsertionPoint] type. The methods it provides are meant
+ * to be used by the authors of custom insertion points.
+ */
+@Internal
+public sealed interface CoordinatesFactory {
+
+    /**
+     * Creates coordinates pointing at a specific line and column in the file.
+     */
+    public fun at(line: Int, column: Int): TextCoordinates {
+        val cursor = Cursor.newBuilder()
+            .setLine(line)
+            .setColumn(column)
+        return TextCoordinates.newBuilder()
+            .setInline(Position.newBuilder().setCursor(cursor))
+            .build()
+    }
+
+    /**
+     * Creates coordinates pointing at the beginning of a specific line in the text.
+     */
+    public fun atLine(line: Int): TextCoordinates =
+        TextCoordinates.newBuilder()
+            .setWholeLine(line)
+            .build()
+
+    /**
+     * Creates coordinates pointing at the beginning of the first line in the text.
+     */
+    public fun startOfFile(): TextCoordinates =
+        atLine(0)
+
+    /**
+     * Creates coordinates pointing at the point after the last line in the text.
+     */
+    public fun endOfFile(): TextCoordinates =
+        TextCoordinates.newBuilder()
+            .setEndOfFile(Empty.getDefaultInstance())
+            .build()
+
+    /**
+     * Creates coordinates that do not point at anywhere in the text.
```

Comment:
This is a bit controversial comment for a thing called `InsertionPoint`. Where are we going to insert, if at all then? Please elaborate.